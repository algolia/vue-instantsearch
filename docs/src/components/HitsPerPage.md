---
title: HitsPerPage
mainTitle: Components
layout: main.pug
category: Components
withHeadings: true
navWeight: 6
editable: true
githubSource: docs/src/components/HitsPerPage.md
---

Displays a `select` element allowing users to choose the number of hits to display.

<a class="btn btn-static-theme" href="stories/?selectedKind=HitsPerPage">ðŸ•¹ try out live</a>

## Usage

```html
<ais-hits-per-page
  :items="[{
    label: '20 results',
    value: 20,
    default
  }, {
    label: '100 results',
    value: 100
  }]"
  ></ais-hits-per-page>
```

## Props

Name | Type | Required | Default | Description | 
---|---|---|---|---
items | `[<Object>, ...]` | yes | no | Array of possible values for the select, example: <br/>`[{label: 'Ten results', value: 10, default}, {label: 'Twenty results', value: 20}]`.<br/>The item with `default: true` will be the one applied by default.

## CSS classes

Here's a list of CSS classes exposed by this widget. To better understand the underlying
DOM structure, have a look at the generated DOM in your browser.

Class name | Description
---|---
`ais-HitsPerPage` | Container class  
`ais-HitsPerPage-select` | A refinement option
