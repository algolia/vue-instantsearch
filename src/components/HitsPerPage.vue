<template>
  <div
    v-if="state"
    :class="suit()"
  >
    <slot
      :items="state.items"
      :refine="state.refine"
      :hasNoResults="state.hasNoResults"
    >
      <select
        :class="suit('select')"
        v-model="selected"
        @change="handleChange"
      >
        <option
          v-for="item in state.items"
          :key="item.value"
          :class="suit('option')"
          :value="item.value"
        >{{ item.label }}</option>
      </select>
    </slot>
  </div>
</template>

<script>
import { connectHitsPerPage } from 'instantsearch.js/es/connectors';
import { createPanelConsumerMixin } from '../mixins/panel';
import { createWidgetMixin } from '../mixins/widget';
import { createSuitMixin } from '../mixins/suit';

export default {
  name: 'AisHitsPerPage',
  mixins: [
    createSuitMixin({ name: 'HitsPerPage' }),
    createWidgetMixin({ connector: connectHitsPerPage }),
    createPanelConsumerMixin({
      mapStateToCanRefine: state => state.hasNoResults === false,
    }),
  ],
  props: {
    items: {
      type: Array,
      required: true,
      default: () => [],
    },
    transformItems: {
      type: Function,
      default(items) {
        return items;
      },
    },
  },
  data() {
    return {
      selected: this.items.find(item => item.default === true).value,
    };
  },
  computed: {
    widgetParams() {
      return {
        items: this.items,
        transformItems: this.transformItems,
      };
    },
  },
  methods: {
    handleChange() {
      this.state.refine(this.selected);
    },
  },
};
</script>
