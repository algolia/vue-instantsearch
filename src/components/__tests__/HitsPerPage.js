import { mount } from '@vue/test-utils';

import HitsPerPage from '../HitsPerPage.vue';
jest.mock('../../component');

it('renders correctly', () => {
  const wrapper = mount(HitsPerPage, {
    propsData: {
      items: [
        {
          label: '10 results',
          value: 10,
          default: true,
        },
        {
          label: '20 results',
          value: 20,
        },
      ],
    },
  });
  expect(wrapper.html()).toMatchSnapshot();
});

it('behaves correctly', () => {
  const wrapper = mount(HitsPerPage, {
    propsData: {
      items: [
        {
          label: '10 results',
          value: 10,
          default: true,
        },
        {
          label: '20 results',
          value: 20,
        },
      ],
    },
  });
  // This is bad 👇🏽 but the only way for now to trigger changes
  // on a select: https://github.com/vuejs/vue-test-utils/issues/260
  wrapper.vm.selected = 20;
  wrapper.find('select').trigger('change');
  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith(20);
});
