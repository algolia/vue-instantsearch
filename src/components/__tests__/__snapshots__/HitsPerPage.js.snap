// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `

<div class="ais-HitsPerPage">
  <select class="ais-HitsPerPage-select">
    <option class="ais-HitsPerPage-option"
            value="10"
    >
      10 results
    </option>
    <option class="ais-HitsPerPage-option"
            value="20"
    >
      20 results
    </option>
  </select>
</div>

`;
